### PR TITLE
Dataset improvements

### PR DESCRIPTION
**Pull request by Sauro98**
Hi, I'm opening the PR a little early because I'd like to ask some questions. This PR wants to make the changes proposed in #70.

* struct `Dataset` has been renamed `DatasetBase`
* All crates have been updated according to the renaming of `Dataset` into `DatasetBase`
* New types have been added:
    * `Dataset`: owned `Array2` as records, owned `Array1` as targets. The current implementation allows to write `Dataset<f24,f24>` or `Dataset<f24,usize>` and so on. Is this in line with the objective of the proposed changes?
    * `DatasetView`: view of an `Array2` as records, view of an `Array1` as targets.  The current implementation allows to write `DatasetView<f24,f24>` or `DatasetView<f24,usize>` and so on. Calling `view()` on a `Dataset` produces a `DatasetView`.
 * Functions inside sub-crate `datasets` have been changed to return an instance of `Dataset`.

There's still k-folding missing but I figured that it would be better to have an intermediate check. I also added tests to ensure the  existence of the implementations of the required methods for the two new types and they also provide a look at how the two types can be used

### CODE REVIEW

File: src/dataset/mod.rs [hunk 1]
```diff
@@ -96,3 +108,144 @@ where
...
+
+        let linspace: Array1<f64> = Array1::linspace(0.0, 0.8, 100);
+        let records = Array2::from_shape_vec((50, 2), linspace.to_vec()).unwrap();
+        let targets: Array1<f64> = Array1::linspace(0.0, 0.8, 50);
+        let dataset = Dataset::new(records, targets);
+
+        //Split with ratio view
+        let (train, val) = dataset.split_with_ratio_view(0.5);
+        assert_eq!(train.targets().len(), 25);
+        assert_eq!(val.targets().len(), 25);
```
Comment (Sauro98) [inline src/dataset/mod.rs:146]: `split_with_ratio_view` for type `Dataset` uses the already existing implementation of the method, but this returns a slice as the targets. Would it be better to re-implement it for `Dataset` so that it returns an `ArrayView1` as the targets?

File: src/dataset/impl_dataset.rs [hunk 2]
```diff
@@ -247,13 +247,122 @@ impl<F: Float, D: Data<Elem = F>, I: Dimension> From<ArrayBase<D, I>>
...
+
+        let new_targets = targets.iter().map(fnc).collect::<Vec<T>>();
+        let new_targets = Array1::from_shape_vec(new_targets.len(), new_targets).unwrap();
+
+        DatasetBase {
+            records,
+            targets: new_targets,
+            weights,
+        }
+    }
```
Comment (Sauro98) [inline src/dataset/impl_dataset.rs:278]: I had to add it in order to change the return type of functions in the `datasets` sub-crate, since `map_targets` puts the new targets in a `Vec`

**Comment by bytesnake** [comment]
awesome! I'm currently traveling and will be back end of december, but the changes are looking good :+1: for `map_targets`: just drop the Vec implementation, it is only confusing for the user

20.12.2020 10:59:02 Ivano Donadi <notifications@github.com>:

**Comment by bytesnake** [comment]
yes!

20.12.2020 10:56:43 Ivano Donadi <notifications@github.com>:

**Comment by Sauro98** [comment]
Hi, I've added k-folding for `Dataset` and `DatasetView` types, implemented the changes mentioned in the previous comments and I've added a shy attempt at documentation. There are some questions I would like to ask:
*  Is that style/content of documentation alright? If so, or even if you have some suggestions, I would gladly write some more for the `impl_dataset.rs` file
* I see that in general the methods in this file don't check for correctness of input parameters, and so I left the `k` parameter for k-folding unchecked. Is this alright? Or else, should methods instead return a `Result` catching any error with the inputs?

Of course, I'm here for any comments about the changes already implemented. Happy holidays ☃️

**Comment by bytesnake** [comment]
> Hi, I've added k-folding for `Dataset` and `DatasetView` types, implemented the changes mentioned in the previous comments and I've added a shy attempt at documentation. There are some questions I would like to ask:
> 
>     * Is that style/content of documentation alright? If so, or even if you have some suggestions, I would gladly write some more for the `impl_dataset.rs` file
> 

They are looking good, I think that documentation should really be improved. You can take this blogpost as a guide how to document functions: https://deterministic.space/machine-readable-inline-markdown-code-cocumentation.html

>     * I see that in general the methods in this file don't check for correctness of input parameters, and so I left the `k` parameter for k-folding unchecked. Is this alright? Or else, should methods instead return a `Result` catching any error with the inputs?

yes we should, but only once we have proper error handling with `thiserror`

> 
> Of course, I'm here for any comments about the changes already implemented. Happy holidays snowman_with_snow

thanks :)

File: src/dataset/mod.rs [hunk 3]
```diff
@@ -70,6 +70,18 @@ where
     weights: Vec<f32>,
 }
 
+/// Dataset
+///
+/// The most commonly used typed of dataset. It contains a number of records
+/// stored as an `Array2` and each record corresponds to a single target. Such
+/// targets are stored as an `Array1`.
+pub type Dataset<D, T> = DatasetBase<ArrayBase<OwnedRepr<D>, Ix2>, ArrayBase<OwnedRepr<T>, Ix1>>;
```
Comment (bytesnake) [inline src/dataset/mod.rs:78]: should we have a `type DatasetPr<D> = DatasetBase<ArrayBase<OwnedRepr<D>, Ix2>, ArrayBase<OwnedRepr<Pr>, Ix2>>` as well to multiclass probabilities :thinking:

**Review by bytesnake — COMMENTED** [review]
thanks for 
 * switching to `DatasetBase`
 *  extending testing and documentation
 * introducing k-folding

**Comment by bytesnake** [comment]
I merged the naive bayes PR, can you rebase?

**Comment by bytesnake** [comment]
one idea which would improve readability:
add a field to `DatasetBase` called `features: Vec<String>` to give each feature a name. This could be used in printing correlation or decision tree etc.
But this would probably go into a follow up PR as these are already enough changes

File: linfa-clustering/benches/gaussian_mixture.rs [hunk 4]
```diff
@@ -22,7 +22,7 @@ fn gaussian_mixture_bench(c: &mut Criterion) {
             let n_features = 3;
             let centroids =
                 Array2::random_using((n_clusters, n_features), Uniform::new(-30., 30.), &mut rng);
-            let dataset = Dataset::from(generate_blobs(cluster_size, &centroids, &mut rng));
+            let dataset = DatasetBase::from(generate_blobs(cluster_size, &centroids, &mut rng));
```
[the remark this answers is not in the document]
Comment (Sauro98) [reply linfa-clustering/benches/gaussian_mixture.rs:25]: `Dataset<_,_>` at the moment does not allow for `()` targets and so `generate_blobs` can't be cast to a dataset, unless it is coupled with some kind of empty array

File: src/dataset/mod.rs (hunk 3 above)
Comment (Sauro98) [reply src/dataset/mod.rs:78]: I think that would be a handy shortcut

**Review by Sauro98 — COMMENTED** [review]
did not mean to start a review

File: src/dataset/impl_dataset.rs [hunk 5]
```diff
@@ -247,13 +247,237 @@ impl<F: Float, D: Data<Elem = F>, I: Dimension> From<ArrayBase<D, I>>
...
+                Dataset::new(remaining_records, remaining_targets),
+                // validation
+                Dataset::new(folds_records[0].into_owned(), folds_targets[0].into_owned()),
+            ));
+
+            folds_records.rotate_left(1);
+            folds_targets.rotate_left(1);
+        }
+        res
+    }
```
Comment (Sauro98) [inline src/dataset/impl_dataset.rs:482]: How does this look? Is there a better way than shifting the vectors at each iteration to generate all permutations?

Comment (Sauro98) [reply src/dataset/impl_dataset.rs:482]: Maybe collecting into an `Array1` and then using `select` on the chunks indices?

File: linfa-clustering/benches/gaussian_mixture.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bytesnake) [reply linfa-clustering/benches/gaussian_mixture.rs]: that's true

File: src/dataset/impl_dataset.rs (hunk 5 above)
Comment (bytesnake) [reply src/dataset/impl_dataset.rs:482]: yes much better :+1: I did not anticipate that you cannot concatenate views in ndarray
perhaps we can do this without copying by:
 * changing to `fn fold(&'a mut self, k: usize) -> impl Iterator<Item = (Dataset<_, _>, Dataset<_, _>)>`
 * destructing the arrays with `as_slice` and swapping the validation set into place with https://doc.rust-lang.org/stable/std/primitive.slice.html#method.swap_with_slice
 *  return a view with `ArrayView::from_shape`

this would borrow the dataset for the k-folding and return the dataset in original order

Comment (bytesnake) [reply src/dataset/impl_dataset.rs:482]: perhaps we should name that `iter_fold` and add later on

File: src/dataset/impl_dataset.rs [hunk 6]
```diff
@@ -43,47 +47,53 @@ impl<R: Records, S: Targets> Dataset<R, S> {
...
+        DatasetBase {
             records,
             targets: new_targets,
             weights,
         }
     }
 }
 
-impl<F: Float, T: Clone> Dataset<Array2<F>, Vec<T>> {
+impl<F: Float, T: Clone> DatasetBase<Array2<F>, Vec<T>> {
```
Comment (bytesnake) [inline src/dataset/impl_dataset.rs:96]: can't you drop this implementation?

File: src/dataset/impl_dataset.rs (hunk 5 above)
Comment (Sauro98) [reply src/dataset/impl_dataset.rs:482]: Hi, I have a couple of doubts. Please bear with me as this is the first time I've tried to do something like this:

* The way I've understood it is: take a slice for the first fold and a slice for the rest of the array, then, at every step of the iterator, swap the content of the fold with the content of the remaining array at the position of the next fold. Is that what you meant?
* Wouldn't `swap_with_slice` panic if `k` is such that the last chunk does not have the same length as the others?

Comment (bytesnake) [reply src/dataset/impl_dataset.rs:482]: > Hi, I have a couple of doubts. Please bear with me as this is the first time I've tried to do something like this:
> 
>     * The way I've understood it is: take a slice for the first fold and a slice for the rest of the array, then, at every step of the iterator, swap the content of the fold with the content of the remaining array at the position of the next fold. Is that what you meant?
yes and of course swap the content back in the next step otherwise you would change order of chunks

> 
>     * Wouldn't `swap_with_slice` panic if `k` is such that the last chunk does not have the same length as the others?
I would propose to ignore the trail of the dataset and always include it in the training set

> 
>     * Doesn't `swap_with_slice` want a mutable slice?  Can I take a mutable slice from an `ArrayView`? Shouldn't it be only possible for an `ArrayViewMut`?

you should be able to create a mutable view of the underlying arrays with `pub fn view_mut(&mut self) -> ArrayViewMut<'_, A, D> `

Comment (bytesnake) [reply src/dataset/impl_dataset.rs:482]: are you in the middle of creating this or should we merge the PR and amend such a function in a follow-up?

Comment (Sauro98) [reply src/dataset/impl_dataset.rs:482]: > you should be able to create a mutable view of the underlying arrays with pub fn view_mut(&mut self) -> ArrayViewMut<'_, A, D>
So I can only implement it for `Dataset` and not `DatasetView` right? Because I can't take a mutable view of an `ArrayView`.

I'm currently trying to implement it but If you say so we can stop here for this PR and implement it another time.  I still have a commit to push with the changes you suggested. For `fold` i substitute the shifting of the `Vec` with just a swap of the first chunk with the one in the next fold.

Comment (Sauro98) [reply src/dataset/impl_dataset.rs]: I've tried my hand at it and I'm having some issues with lifetimes, I think it would be better to leave it for an eventual follow-up PR

Comment (bytesnake) [reply src/dataset/impl_dataset.rs]: :+1:

**Review by bytesnake — APPROVED** [review]
